### DISCUSSION TITLE
How to distribute plugins?

### DISCUSSION THREAD

**Discussion by Timmmm** [pawelsalawa/sqlitestudio#3922] [Q&A]
I have written a custom plugin similar to `MultiEditorImage`, called `MultiEditorVarints` (it decodes a `BLOB` of varints). I put it next to the `MultiEditorImage` in the source, like this:

```
![image](https://user-images.githubusercontent.com/376842/105727689-3d304600-5f23-11eb-95ea-f64e686cff62.png)
```

I can build it and it runs fine. But I'd like to distribute it. It generates `libMultiEditorVarints.dylib` file, and I've tried copying that into the `/Applications/SQLIteStudio.app/Contents/PlugIns` directory of the current release version. However when I start that version it gives this error:

> Data editor plugin '%1' not loaded, while it is defined for editing '%1' data type.

![image](https://user-images.githubusercontent.com/376842/105727938-7f598780-5f23-11eb-95b8-a595e3397514.png)

(I think there's a missing `.arg()` in there somewhere!)

If I go into the Plugins section of Preferences my plugin is not there. Should this work? Am I missing a step?

I'm also compiling my plugin from today's `master`, but copying it to the 3.2.1 release - would you expect that to break it? Is distributing plugins even supported?

Any help appreciated!

**Comment by pawelsalawa** [comment]
I think it might be related to dependency resolving.
When you put your plugin to the `PlugIns` directory, please go there within your terminal command line and type `otool -L libMultiEditorVarints.dylib`. You can compare it with result for any other plugin in this directory. Seeing anything unusual?

**Reply by Timmmm** [reply]
For the current release:

```
libMultiEditorImage.dylib:
	libMultiEditorImage.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcoreSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libguiSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.10.0, current version 5.10.0)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.10.0, current version 5.10.0)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.10.0, current version 5.10.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

Built from source:

```
libMultiEditorImage.dylib:
	libMultiEditorImage.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcoreSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libguiSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.10.0, current version 5.10.0)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.10.0, current version 5.10.0)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.10.0, current version 5.10.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

That appears to be identical. Here's my plugin, which is pretty similar:

```
libMultiEditorVarints.dylib:
	libMultiEditorVarints.dylib (compatibility version 0.0.0, current version 0.0.0)
	libcoreSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	libguiSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.15.0, current version 5.15.2)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.15.0, current version 5.15.2)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2022.20.117)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 244.32.7)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.15.0, current version 5.15.2)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

Except hmm it doesn't seem to search the rpath for `libcoreSQLiteStudio.1.dylib` or `libguiSQLIteStudio.1.dylib`. Is there some build step that fixes plugins up that I need to modify? And I don't know why it needs AppKit or Metal, when MultiEditorImage doesn't.

I pretty much just copied the `MultiEditorImage` directory to `MultiEditorVarints`, then renamed every instance of `MultiEditorImage` to `MultiEditorVarints`. Hmm.

**Reply by Timmmm** [reply]
Oh you know what, I just noticed I didn't actually copy the `package.xml` file and it seems like that is used in the `Fixing paths in ...` bundle step which sounds like it is an rpath fix! Oops!

**Reply by Timmmm** [reply]
Damn, that did fix the rpath issue (though it still weirdly depends on AppKit and Metal, whereas MultiEditorImage doesn't):

```
ibMultiEditorVarints.dylib:
        libMultiEditorVarints.dylib (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libcoreSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
        @rpath/libguiSQLiteStudio.1.dylib (compatibility version 1.0.0, current version 1.0.0)
        @rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.15.0, current version 5.15.2)
        @rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.15.0, current version 5.15.2)
        /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2022.20.117)
        /System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 244.32.7)
        @rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.15.0, current version 5.15.2)
        /System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
        /System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
        /System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
        /System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

But unfortunately I'm still getting the same error as before:

    [15:07:04] Data editor plugin '%1' not loaded, while it is defined for editing '%1' data type.

**Reply by pawelsalawa** [reply]
How about you try to run the app from command line with options: `-d --debug-stdout` and see if it prints anything useful.

**Reply by pawelsalawa** [reply]
I just noticed you're using master branch. It's a really bad idea right now. It's a mess, waiting for fix.

Best what you can do is to download source code package for 3.2.1 - it is available at Releases page: https://github.com/pawelsalawa/sqlitestudio/archive/3.2.1.zip

**Reply by Timmmm** [reply]
Ok I tried this again with the 3.3.2 release and it works now! I think the install name handling is still pretty janky. For some reason when you do `make bundle` it unfixes the install names in the plugins... or something. Hard to follow anyway.

Rather than compiling everything and fixing the install names afterwards it would be better to specify them during compilation. That would also mean you don't need to fix the plugin install names at all because they would inherit the right names in the first place.

E.g. in the `macx` section of `coreSQLiteStudio.pro` maybe try adding something like this:

```
    QMAKE_LFLAGS_SONAME = -Wl,-install_name,@rpath/$${TARGET}.$${QMAKE_EXTENSION_SHLIB}
```

Anyway thanks for doing the 3.3 release! Dark mode is great.

**Reply by pawelsalawa** [reply]
Are you sure you're compiling your plugin with the same Qt that SQLiteStudio is compiled with? Otherwise your plugin would refer to symbol addresses that do not exist in the Qt shipped with SQLiteStudio.

**Reply by pawelsalawa** [reply]
Oh, I just re-read your message and you said it worked this time. Good.

Regarding compliation flags - I would really rather not change anything. I spent tons of hours to get macosx boundle built properly and I don't to mess it up.
